### PR TITLE
tests/raftstore: refact call_command_on_leader

### PR DESCRIPTION
**Pull request by tiancaiamao**
auto retry if response has error and the error is caused
by leader change or split

### CODE REVIEW

File: tests/raftstore/cluster.rs [hunk 1]
```diff
@@ -146,19 +146,45 @@ impl<T: Simulator> Cluster<T> {
...
                                   timeout: Duration)
                                   -> Result<RaftCmdResponse> {
-        for _ in 0..200 {
-            if let Some(leader) = self.leader_of_region(region_id) {
-                request.mut_header().set_peer(leader);
-                return self.call_command(request, timeout);
-            }
-            sleep_ms(10);
+        let mut retry_cnt = 0;
+        loop {
```
Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: loop ... match ... continue ... return ...
I feel a little dizzy...
PTAL @BusyJay

File: tests/raftstore/cluster.rs [hunk 2]
```diff
@@ -146,19 +146,45 @@ impl<T: Simulator> Cluster<T> {
...
+            let result = self.naive_call_command_on_leader(region_id, request.clone(), timeout);
+            match result {
+                Ok(resp) => {
```
Comment (ngaut) [inline tests/raftstore/cluster.rs]: Use if let pattern to simplify code.

Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: I tried to use if let, but always get a partial moved error  -_-!

File: tests/raftstore/cluster.rs [hunk 3]
```diff
@@ -146,19 +146,42 @@ impl<T: Simulator> Cluster<T> {
...
+        let mut retry_cnt = 0;
+        loop {
+            let result = self.naive_call_command_on_leader(region_id, request.clone(), timeout);
+            if let Ok(resp) = result {
+                if resp.get_header().has_error() && retry_cnt < 10 {
+                    retry_cnt += 1;
+                    if self.refresh_leader_if_needed(&resp, region_id) {
+                        warn!("seems leader changed, let's retry");
+                        continue;
+                    } else if resp.get_header().get_error().has_stale_epoch() {
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: stale epoch checking is only ok for common request like get/put, not for admin command,

File: tests/raftstore/cluster.rs [hunk 4]
```diff
@@ -146,19 +146,42 @@ impl<T: Simulator> Cluster<T> {
...
-                                  mut request: RaftCmdRequest,
+                                  request: RaftCmdRequest,
                                   timeout: Duration)
                                   -> Result<RaftCmdResponse> {
-        for _ in 0..200 {
-            if let Some(leader) = self.leader_of_region(region_id) {
-                request.mut_header().set_peer(leader);
-                return self.call_command(request, timeout);
+        let mut retry_cnt = 0;
+        loop {
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: you can use `for _ in 0..retry_cnt` directly.

File: tests/raftstore/cluster.rs [hunk 5]
```diff
@@ -146,19 +146,42 @@ impl<T: Simulator> Cluster<T> {
         self.sim.rl().call_command(request, timeout)
     }
 
+    fn naive_call_command_on_leader(&mut self,
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: naive confuses me, why naive? 

I think you can move this to call_command_on_leader directly.

File: tests/raftstore/cluster.rs (hunk 4 above)
Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: for _ in 0..retry_cnt {
            let result = ...
            if let Ok(_) = result {
                continue
            } else {
                return ...
            }
        }
        // how to return result after retry loop finish

File: tests/raftstore/cluster.rs [hunk 6]
```diff
@@ -146,19 +146,39 @@ impl<T: Simulator> Cluster<T> {
         self.sim.rl().call_command(request, timeout)
     }
 
+    fn call_command_on_leader_directly(&mut self,
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: /directly/once

**Comment by siddontang** [comment]
LGTM

File: tests/raftstore/cluster.rs [hunk 7]
```diff
@@ -146,19 +146,39 @@ impl<T: Simulator> Cluster<T> {
...
                                   timeout: Duration)
                                   -> Result<RaftCmdResponse> {
-        for _ in 0..200 {
-            if let Some(leader) = self.leader_of_region(region_id) {
-                request.mut_header().set_peer(leader);
-                return self.call_command(request, timeout);
+        let mut retry_cnt = 0;
+        loop {
+            let result = self.call_command_on_leader_once(region_id, request.clone(), timeout);
+            if let Ok(resp) = result {
```
Comment (BusyJay) [inline tests/raftstore/cluster.rs]: Matching err first can reduce indention. If using `if let` with else, I personally prefer just using `match`.

File: tests/raftstore/cluster.rs (hunk 4 above)
Comment (BusyJay) [inline tests/raftstore/cluster.rs]: You can define result outside loop.

File: tests/raftstore/cluster.rs [hunk 8]
```diff
@@ -146,19 +146,39 @@ impl<T: Simulator> Cluster<T> {
...
+                if resp.get_header().has_error() && retry_cnt < 10 {
```
Comment (BusyJay) [inline tests/raftstore/cluster.rs]: `refresh_leader_if_needed` will check if response has error, so no need check here.

File: tests/raftstore/cluster.rs (hunk 4 above)
Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: let mut result: Result<RaftCmdResponse>;
        for _ in 0..retry_cnt {
                result = ...;
                if let Ok(_) = result {
                        coutinue
                }
        }
        result

compiler would complain "use of partially moved value" of result

File: tests/raftstore/cluster.rs (hunk 7 above)
Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: previously `match` was used, and changed to `if let`which was suggested by @ngaut 
...now, I've give up trying to make it clearer after several attempts

Comment (BusyJay) [inline tests/raftstore/cluster.rs]: What I mean is to use `if result.is_err() { return result; }`.

Comment (siddontang) [inline tests/raftstore/cluster.rs]: I prefer match than `if let` + `else` too. 

If we want to handle only one case, `if let` is better.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**